DOC/TST: (nearly) finalize current use of `xp_capabilities` and its documentation

steppi opened the pull request:
#### Reference issue

Closes #23582

Closes #24026

#### What does this implement/fix?

This PR does a number of things which I believe brings the project to add ``xp_capabilities`` to all functions in SciPy with array API support to completion, verifies that this is the case, and adds comprehensive documentation for the use of `xp_capabilities`. 

##### Strict checking

1. Strict checks have been added that enforce that all tests using the ``xp`` fixture must draw pytest marks from `xp_capabilities` using `make_xp_test_case` or one of its friends, unless the test is explicitly marked as not using `xp_capabilities` with the marker `pytest.mark.uses_xp_capabilities(False)`. The way this works is that `make_xp_pytest_marks` inserts the marker `uses_xp_capabilities(True)`, and the `xp` fixture has been updated to enforce that the `uses_xp_capabilities` mark exists.
2. A tool, `scipy/tools/check_xp_untested.py` has been added, which can be called with `spin check-xp-untested`, which checks for functions (methods/classes) with `xp_capabilities` entries advertising support for an alternative backend for which there are no associated tests using the `xp` fixture and `make_xp_test_case` or one of its friends. This is now run in CI in the lint jobs, enforcing that any advertisement of alternative backend support must come with tests.

Taken together, this means we can confirm that `xp_capabilities` has been added for all functions in the public API which have some level of alternative backend support. There may be functions which incidentally work with one or more alternative backends, but since they are not tested, I think it's fair that they are documented as unsupported. PRs like https://github.com/scipy/scipy/pull/24128 adding tests for such functions are highly welcome. This also means that we are not in danger of falling behind in adding `xp_capabilities` when support is added for additional functions. CI will enforce that `xp_capabilities` is used.

The use of these strict checks let me quickly round up all remaining cases where `xp_capabilities` didn't exist or
weren't being used, and some cases where they weren't correct (when advertising support without tests). That's part of the reason why this PR touches so many files. It was also necessary to add `pytest.mark.uses_xp_capabilities(False)` to tests for private functionality for which there are no relevant `xp_capabilities` entries.

##### Comprehensive documentation for the use of `xp_capabilities`

I've also added comprehensive documentation for the use of `xp_capabilities` based on what I've learned over the past few months helping to add `xp_capabilities` everywhere and improve testing.

 In at least one case, a recommendation I had made previously has been rethought and changed. Before I was really hung up on having `SCIPY_DEFAULT_DTYPE` control the precision of inputs in expected outputs in tests, but I realized that this is actually kind of dumb, so sorry for creating noise around that before. Really, the purpose of `SCIPY_DEFAULT_DTYPE=float32` should be to test for subtle bugs that can crop up with the `torch` backend due to the use of array creation functions without explicit dtypes within functions. I realized, that really, it makes sense that the input dtypes and expected output dtypes in tests should be controlled mindfully, and should not be determined by `SCIPY_DEFAULT_DTYPE`. There are many tests in SciPy which do not follow the practice I now recommend, and I am responsible for many of them. Fortunately, this is something that I don't anticipate being difficult to hammer out over time. By changing this recommendation, I was able to remove some quite ugly recommended examples for ensuring test inputs have the default dtype. The ugliness of these examples should have been a clue that something was amiss.

I hope all of the other recommendations I've made in `scipy/docs/source/dev/api-dev/array_api.rst` are not surprising and not controversial. I just tried to describe the practice I worked towards over the past few months. As a follow-up I want to help finish up https://github.com/scipy/scipy/pull/23960 and add some info to the documentation about caveats to watch out for when using `xp_capabilities` with classes.

#### Additional information
<!--Any additional information you think is important.-->
I've observed some failures locally in some `stats` tests that use the `xp` fixture, but they also happen for me on `main` so I'm not sure if they will crop up in CI here. I'm not sure if the checks for untested functions with advertised backend support will actually work in CI, but if errors pop up because I messed up something in `.github/workflows/lint.yml`, it will likely not be difficult to fix. I did not use AI tools to produce the code here, but as I first pass I did try to get Gemini to implement the strict checks I wanted. It produced some horrible complicated thing that parsed abstract syntax trees. What I came up with was an attempt at a simple-minded but robust [grug-brained](https://grugbrain.dev/) solution, that I made while recoiling in horror at what Gemini produced. I'm aware that directly using Gemini's solution would likely have been iffy from a copyright perspective.

steppi committed:
TST: Add marker for asserting test uses xp_capabilities

steppi committed:
TST: Make tests pass that fail because xp fixture no xp_capabilities

steppi committed:
TST: Check in work in progress, check xp_cap no tests

steppi committed:
TST: Make check_xp_untested a standalone tool

steppi committed:
DEV: Add spin command for check_xp_untested

steppi committed:
TST: explain if check_xp_untested run without SCIPY_ARRAY_API=1

steppi committed:
CI: Add check-xp-untested to CI

steppi committed:
MAINT: fix bug in marker collection

steppi committed:
TST: Fix some signal xp_capabilities

steppi committed:
TST/DOC: some xp_capability/pytest mark fixes for ndimage

steppi committed:
TST/DOC: mark PchipInterpolator as np_only since not tested

steppi committed:
TST: Add test for ttest_rel using xp fixture

steppi committed:
TST/DOC: some spatial xp_capabilities/pytest mark adjustments

steppi committed:
TST: Add tests for cluster linkage wrappers

steppi committed:
DOC/TST: correctly mark cubature rules as private why no xp_cap

steppi committed:
DOC: Add some nuance to reason ``vq`` not supported on GPU

steppi committed:
DOC: very minor single word change for better flow

steppi committed:
DOC: Add note on dispatch to native implementations in array api docs

steppi committed:
DOC: Check in current state of array api standard support docs

steppi committed:
DOC: Add section on unsupported function and move out-of-scope part

steppi committed:
DOC: More doc updates

steppi committed:
DOC: minor tweak

steppi committed:
DOC: Add documentations on xp_cap skip/xfail_backends

steppi committed:
DOC: add documentation for ``extra_note`` xp_cap

steppi committed:
DOC: Add warning about docstring overwriting breaking references

steppi committed:
DOC: more docs on adding tests

steppi committed:
DOC: update no xp_cap reasons

steppi committed:
DOC: more testing documentation

steppi committed:
DOC: even more array api testing documentation

steppi committed:
DOC: more testing docs

steppi committed:
DOC: tweak wording describing SCIPY_DEFAULT_DTYPE

steppi committed:
DOC: update backend isolation section to reflect new thinking

steppi committed:
DOC: add link for backend isolation and adjust text width

steppi committed:
DOC: update section on testing with jax jit

steppi committed:
DOC: more updates for lazy testing docs

steppi committed:
DOC: explain that tests are not optional

steppi committed:
DEV: Make check_xp_untested check for np_only with no exceptions

- not just np_only=True

steppi committed:
DOC/TST: Get rid of exception for signal.czt since untested

steppi committed:
DOC: simplifications to table code, don't track missing xp_caps

steppi committed:
DOC: Tweaks for array api testing docs

steppi committed:
STY: appease the linter

mdhaber commented on scipy/stats/tests/test_continued_fraction.py:
```diff
@@ -14,6 +14,7 @@
 # n = int(xp.real(xp_ravel(n))[0])
 # (at some point in here the shape becomes nan)
 @pytest.mark.skip_xp_backends('dask.array', reason="dask has issues with the shapes")
+@pytest.mark.uses_xp_capabilities(False, reason="private")
```
I think it would be about as easy to use `xp_capabilities`, unless private functions are not supposed to.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -5037,6 +5039,31 @@ def test_location_sign(self, ksfunc, alternative,
+    # stats.ttest_rel had no tests using the xp fixture, which causes a failure
+    # in tools/check_xp_untested.py. A portion of the below test has been
+    # converted as a temporarily measure. TODO: Convert all of test_ttest_rel
+    # and refactor it into separate tests.
```
That's because the entire implementation of `ttest_rel` is:
```python3
    return ttest_1samp(a - b, popmean=0., axis=axis, alternative=alternative,
                       _no_deco=True)
```
It's basically just an alias now.
So if `ttest_1samp` is adequately tested, so is `ttest_rel`. 
I think it's worth skimming the `ttest_rel` tests for unique ideas and adding any to the `ttest_1samp` tests, but I think we can let go of the rest.

Then, if needed, I think we could run all the `ttest_1samp` on a wrapper function that calls `ttest_rel` because - numerical precision details aside - tests of these three things are equivalent: 
```python3
ttest_1samp(x, popmean)
ttest_1samp(x - popmean, 0)
ttest_rel(x - popmean + y, y)  # y is random of the same shape of x
```

mdhaber reviewed (commented):
Mostly some thoughts about the stats stuff. It's too late to really look at the `rst` tonight.

j-bowhay commented on scipy/cluster/tests/test_hierarchy.py:
```diff
@@ -1205,6 +1224,7 @@ def test_optimal_leaf_ordering(xp):
 @skip_xp_backends(np_only=True, reason='`Heap` only supports NumPy backend')
+@pytest.mark.uses_xp_capabilities(False, reason="private")
 def test_Heap(xp):
     values = xp.asarray([2, -1, 0, -1.5, 3])
```
Maybe here is it cleaner just to remove the `xp` fixture?
```suggestion
def test_Heap():  # `Heap` only supports NumPy backend
    values = np.asarray([2, -1, 0, -1.5, 3])
```

j-bowhay commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3064,6 +3064,7 @@ def test_choose_conv_method_2(xp):
 @skip_xp_backends(np_only=True)
 @pytest.mark.fail_slow(10)
+@pytest.mark.uses_xp_capabilities(False, reason="private")
 def test_filtfilt_gust(xp):
```
```suggestion
@pytest.mark.fail_slow(10)
def test_filtfilt_gust():
```

j-bowhay commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3774,6 +3775,7 @@ def assert_rp_almost_equal(r, p, r_true, p_true, decimal=7):
 
     @skip_xp_backends(np_only=True)
+    @pytest.mark.uses_xp_capabilities(False, reason="private")
     def test_compute_factors(self, xp):
```
```suggestion
    def test_compute_factors(self):
```

lucascolley commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -220,21 +222,380 @@ You would convert this like so::
+
+The ``xp_capabilities`` decorator is what inserts the capabilities table into
+docstrings. It also allows developers to tag tests with ``@make_xp_test_case``
+or ``make_xp_pytest_param`` to automatically generate backend specific
```
nit `@` is in one of these but not the other

steppi replied:
``make_xp_test_case`` is used as a decorator but ``make_xp_pytest_param`` is just a function, so I think the problem here is the way this is worded.

lucascolley commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -220,21 +222,380 @@ You would convert this like so::
+declares that a function works on all supported backends, on JAX with the jit
+and lazily in Dask. This is most likely to occur if a function is written
+entirely in terms of the array API standard as described earlier in this
+document. Such functions are commonly referred to as array API agnostic. For
```
at least by me 😉 feel free to override me empirically

```suggestion
document. Such functions are commonly referred to as array-agnostic. For
```

lucascolley commented on doc/source/dev/api-dev/array_api.rst:
``````````````diff
@@ -270,47 +680,67 @@ which was set by the fixture. Tests without the ``xp`` fixture infer the namespa
+`````````````
+
+After applying these markers, either through ``make_xp_test_case`` and friends, or directly,
+``spin test`` can be used with the new option ``-b`` or ``--array-api-backend``::
``````````````
```suggestion
``spin test`` can be used with the option ``-b`` or ``--array-api-backend``::
```

lucascolley commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -320,110 +750,162 @@ backend, use: ``SCIPY_DEVICE=mps spin test -b torch``.
+``_xp_copy_to_numpy`` in ``scipy._lib._array_api`` which can copy an arbitrary
+``xp`` array to a NumPy array, bypassing any device transfer guards, while
+preserving dtypes. It is essential that this function is only used in
+tests. Attempts to copy a device array to NumPy outside of tests should fail,
```
can we enforce this in CI?

lucascolley commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -320,110 +750,162 @@ backend, use: ``SCIPY_DEVICE=mps spin test -b torch``.
+      # When xp==jax.numpy, _private_toto_helper is wrapped with @jax.jit
+      xp_assert_close(_private_toto_helper(a, b), a)
+
+.. warning::
```
I wonder whether we can reduce duplication between here and array-api-extra by contributing this warning upstream to https://data-apis.org/array-api-extra/generated/array_api_extra.testing.lazy_xp_function.html ? Both places now effectively communicate the problem, but I wonder whether this warning could be shortened here if we add a proper warning admonition upstream.

steppi replied:
Yeah, that seems like a good plan.

lucascolley reviewed (commented):
thanks Albert, fantastic work!! Super happy to see nice docs for this stuff finally. The complexity has got so high that I can no longer rely on my memory to know what is correct/idiomatic!

Feel free to make a best-effort attempt at addressing my comments, there are a lot.

steppi committed:
Apply suggestions from code review [skip ci]

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

steppi commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -220,21 +222,380 @@ You would convert this like so::
+``jax_jit=False``. If a function is neither array API agnostic nor supported
+on JAX through dispatch to a native JAX implementation (again either dispatch to
+a native implementation of the function itself or dispatch for its compiled
+dependencies) then it is virtually certain that the JIT is not supported and
```
Yeah, on a second look I agree. I'll go ahead and remove this part.

steppi commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -220,21 +222,380 @@ You would convert this like so::
+    but does support higher dimensional arrays for ``method="direct"``
+    and ``method="fft"``.
+
+    """
```
This should be changed in `signal/_support_alternative_backends.py` too (but not in this PR).

steppi committed:
Apply suggestions from code review for array_api.rst [skip ci]

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

steppi committed:
Apply suggestions from code review [skip ci]

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

steppi committed:
Update doc/source/dev/api-dev/array_api.rst [skip ci]

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

steppi committed:
Apply review comments

steppi committed:
Undo an accidental formatting change

steppi committed:
Fix bug in check_xp_untested

steppi committed:
Apply more suggestions from review

steppi committed:
DOC: Update outdated note on github actions workflows

steppi committed:
Fix typo in name of new test dependency

steppi commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -220,21 +222,380 @@ You would convert this like so::
+that ``np_only=True`` is only a temporary state. Through the means described
+above in the section on :ref:`adding array API support <dev-arrayapi_adding_support>`,
+it is a reasonable goal for all functions in SciPy's public API to at least reach
+the state ``cpu_only=True``. For functions still waiting in the ``np_only=True`` state,
```
I found a package called [sphinx-togglebutton](https://sphinx-togglebutton.readthedocs.io/en/latest/) that lets us create the expandable admonition. It works pretty well.

steppi committed:
CI: move check-xp-untested to circleci jobs

steppi committed:
CI: move check-xp-untested to own circleci job

steppi committed:
CI: configure check_xp_untested to actually run

steppi committed:
CI: Add test requirements

steppi commented:
The problem with `check_xp_untested` was that the CI job wasn't installing all of the test requirements. Tests missing these optional requirements weren't getting collected, so it looked like there were holes in the test coverage.

steppi committed:
TST: Add some missing uses_xp_capabilities marks

steppi committed:
TST: Do not allow SCIPY_DEFAULT_DTYPE to determine input/output dtypes

steppi commented:
Ha, despite writing this and seemingly having come to my senses:

> I realized, that really, it makes sense that the input dtypes and expected output dtypes in tests should be controlled mindfully, and should not be determined by `SCIPY_DEFAULT_DTYPE`. There are many tests in SciPy which do not follow the practice I now recommend, and I am responsible for many of them.

I couldn't help but add another test that lets `SCIPY_DEFAULT_DTYPE` control the input and output dtypes. That's what that last test failure was about. Fixed now. The other failure on the free-threaded job looks like the classic `differential_evolution`  flakiness.

steppi committed:
Try to get check-xp-untested working with pixi

steppi committed:
Fix some comments in pixi.toml

steppi committed:
Punt on getting check_xp_untested to work with pixi

steppi committed:
Revert accidental whitespace change

steppi committed:
Update pixi.toml

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

steppi committed:
Check in current progress on describing JAX support

DietBru commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -226,15 +228,373 @@ conversion will raise an exception. The reason for that is that silent data
 
+In some cases, compiled code can be supported through delegation to native
+implementations. Such delegation has currently been set up in `scipy.fft`,
+`scipy.ndimage`, `scipy.signal`, and `scipy.special`, though there is not yet a
```
* Writing `~scipy.fft` instead of `scipy.fft` produces the shortened 'fft' (matter of taste)
* There's a [Caveats] section on the "Array API Support: signal" page—why not reference it?

What about someting along the lines of:
```ReST
Such delegation has currently been set up in `~scipy.fft`, `~scipy.ndimage`, `~scipy.signal` (consult :ref:`array_api_support_signal_caveats`), and `~scipy.special` modules,
```

[Caveats]: https://scipy.github.io/devdocs/dev/api-dev/array_api_modules_tables/signal.html#caveats

DietBru reviewed (commented):
It's great to see the documentation improving here. I just left some drive-by comments for `array_api.rst`.

DietBru commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -226,15 +228,373 @@ conversion will raise an exception. The reason for that is that silent data
+      allow_dask_compute=False, jax_jit=True,
+      # Extra note to inject into the docstring
+      extra_note=None,
+  ):
```
I tried that out and it worked as expected. The only issue I see is when writing the docstrs, care must be take with cross-references to other functions...

steppi committed:
Add a note on JAX support

steppi committed:
DOC: More JAX info in docs

steppi committed:
Respond to review comments

steppi committed:
Merge branch 'main' into xp-testing-docs

steppi committed:
Update pythonpath in check-xp-untested job to use py3.13

steppi commented:
> t looks like CircleCI failed because it's using Python 3.11, but SciPy recently dropped support for that version.

I merged main where the python version was updated to Python 3.13 , and updated the Python version in the PYTHONPATH for the `check-xp-untested` job. CircleCI should work now.

steppi committed:
DOC: Add comment discussing alternative way xp_cap might be applied

steppi committed:
DOC: Add recommendation about functions with array inputs

lucascolley reviewed (commented):
thanks Albert, I went back and resolved my previous comments where it looked like they had been addressed. Just a couple more.

ev-br reviewed (commented):
This is a herculean effort @steppi! 

I read through a part of the docs, will need to re-read it again. Several comments below, all optional.

steppi committed:
Apply suggestions from code review [skip ci]

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

steppi committed:
MAINT: Add sphinx-togglebutton to pixi.toml docs dep [skip ci]

steppi committed:
DOC: Get rid of more nonsense I wrote

steppi committed:
DOC: use asarray copy=True instead of xp_copy

steppi committed:
DOC: Revise info on JAX support

steppi committed:
DOC: Add link to JAX gotchas

steppi committed:
DOC: Add section on default dtype

steppi committed:
DOC: break a paragraph up into two smaller ones

steppi committed:
DOC: Replace description of default dtype with link to other part

steppi committed:
EMPTY: [skip ci]

steppi committed:
DOC: Fix a label name [skip ci]

steppi committed:
DOC: Revise default dtype section

steppi committed:
DOC: add mention of type promotion rules and output dtype

steppi committed:
DOC: add examples of using exceptions with xp_capabilities

steppi committed:
DOC: More revisions

steppi committed:
DOC: More revisions

steppi committed:
DOC: More revisions

steppi committed:
DOC: More revisions

steppi committed:
DOC: More revisions

steppi committed:
DOC: More revisions

steppi committed:
DOC: more revisions

steppi committed:
Merge branch 'main' into xp-testing-docs

steppi committed:
MAINT: Update pixi lock file

j-bowhay commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -57,7 +57,8 @@ values:
 CuPy arrays, the expected result for ``vq`` is a ``TypeError``, because ``vq``
-uses compiled code in its implementation, which won't work on GPU.
+uses compiled code in its implementation, which won't work on GPU, and there
+are currently no GPU specific implementations to dispatch to.
```
```suggestion
are currently no GPU specific implementations to delegate to.
```
I was under the impression this was the preferred phrasing

steppi replied:
Indeed. I must have missed this one after the first round of reviews.

steppi committed:
Update doc/source/dev/api-dev/array_api.rst

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

steppi committed:
Update doc/source/dev/api-dev/array_api.rst

j-bowhay commented on scipy/fft/tests/test_real_transforms.py:
```diff
@@ -241,7 +243,16 @@ def test_orthogonalize_dcst3(func, norm, xp):
 
 @skip_xp_backends(np_only=True,
                   reason='array-likes only supported for NumPy backend')
-@pytest.mark.parametrize("func", [dct, idct, dctn, idctn, dst, idst, dstn, idstn])
```
I think we should just remove the `xp` fixture here to save ten lines

steppi replied:
Thanks, good suggestion. I removed `xp` from another test in this file too.

j-bowhay reviewed (approved):
Other than one nit this looks great!

steppi committed:
Merge branch 'main' into xp-testing-docs

steppi committed:
TST: Remove xp fixture from some np_only tests

steppi committed:
DOC: Fix list formatting and mention xp values are raw namespaces

steppi committed:
Merge remote-tracking branch 'origin/xp-testing-docs' into xp-testing-docs

steppi commented:
Thanks @j-bowhay. I've removed the xp fixtures. I also fixed some list formatting in `array_api.rst` and added a note about the `xp` fixture taking values in the set of raw namespaces, not `array_api_compat` wrapped namespaces that @ev-br suggested and I'd forgotten to add.

While working on https://github.com/scipy/scipy/pull/24232, I noticed that `xp_result_type` and `xp_promote` convert an expected integer type to the default floating point type on Torch when `force_floating=True`. This goes counter to what I'm recommending here. I guess mixed integer/float promotion will always be a source of trouble due to it not being defined by the standard. I'm still thinking about what to do about this.

steppi committed:
TST: Fall back to make_xp_test_case

steppi commented:
There's currently not really a nice thing to do in situations like this

```python
 @pytest.mark.parametrize("forward, backward", [make_xp_pytest_param(dct, idct), 
                                                make_xp_pytest_param(dst, idst)]) 
 @pytest.mark.parametrize("type", [1, 2, 3, 4]) 
```

with parametrization over pairs of functions. I just fell back to using `@make_xp_test_case` with four inputs

```python
@make_xp_test_case(dct, idct, dst, idst)
@pytest.mark.parametrize("forward, backward", [(dct, idct), (dst, idst)])
@pytest.mark.parametrize("type", [1, 2, 3, 4])
```
which is fine at least here because all of these functions have the same capabilities.

steppi committed:
Merge branch 'main' into xp-testing-docs

steppi commented:
Thanks @j-bowhay. I think that sounds reasonable, but let's just make sure @ev-br has a chance to review the last round of edits.

tylerjereddy commented on doc/source/dev/api-dev/array_api.rst:
missing a space in `arraysand`

tylerjereddy commented on doc/source/dev/api-dev/array_api.rst:
This also means we basically don't support `torch` with metal performance shader backend on ARM Mac, unless they recently added `float64` (probably not). In practice it kind of works, but our testsuite mostly fails in that scenario last time I checked.

steppi replied:
Yeah. I think there's a path forward to supporting and documenting support for `float32`-only backends in piecemeal fashion, but for the time-being am hoping to limit scope for the first pass of adding array API standard support.

ev-br replied:
> In practice it kind of works, but our testsuite mostly fails in that scenario

Out of curiousity: is it that the code is OK but tests make assumptions which are violated, or is it that the test suite correctly shows that things do not work?

I suppose (pure handwaving) that adding support for these float32-only "metal" backends is a somewhat large project which mostly needs a champion to get started?

Then --- pure handwaving again --- I guess it'd be best to phrase things here to make it clear that both we do not currently support it just yet, and that we're open to evaluating it when somebody makes a case?
Not that I have specific wording suggestions though, so whatever works for Tyler and Albert works for me, too.

rgommers replied:
Note that there's a bunch of discussion on the `float64` topic on the main RFC for array types starting around here: https://github.com/scipy/scipy/issues/18286#issuecomment-2113620407. That still looks relevant and about right.

tylerjereddy reviewed (commented):
Looks complicated, I read through some parts and added only minor comments.

steppi committed:
DOC: Fix typo [docs only]

ev-br reviewed (approved):
All my comments are addressed, and the documentation reads great now. 

> Improving the organization in this way is something I hope we can iterate on.

Definitely agreed! This documentation covers _a lot_ of ground, from basic things all devs should be aware of, to intrinsically tricky esoteria, which hits a small number of us (but when it does, it hits hard indeed), so I think we should merge this herculean effort as is, and iterate as needed.

steppi commented:
Thanks @ev-br!

steppi committed:
DOC: leave more flexibility for float32-only support

steppi committed:
Merge branch 'main' into xp-testing-docs

rgommers committed:
Merge branch 'main' into xp-testing-docs

rgommers commented:
The remaining failures in one job are real, and are a subset (4 out of 5) of the failures reported in gh-24251. They're caused by an unnecessary bump in `pixi.lock` here. That kind of thing is annoying to debug, but I'll emphasize again that no one should be running `pixi update` to bump more than is needed when running `pixi.toml`. xref gh-23637 for improving that.

I'll have a quick look at fixing that up.

rgommers commented:
@steppi why did you add `sphinx-togglebutton`? It looks essentially unmaintained (no release since July 2022) and I'm seeing issues with it locally. We already rely on `sphinx-design` which has dropdowns: https://sphinx-design.readthedocs.io/en/latest/dropdowns.html#dropdowns-in-other-components

steppi committed:
Revert "MAINT: Add sphinx-togglebutton to pixi.toml docs dep [skip ci]"

This reverts commit 662e35074f06230920982db9a1b89fb86c655ff2.

steppi committed:
Revert "MAINT: Update pixi lock file"

This reverts commit 194ab2fde226d25543af35249379bc9d17268c86.

steppi committed:
MAINT: remove sphinx-togglebutton in more places

steppi committed:
DOC: Use sphinx-design for dropdown

steppi committed:
DOC: Hedge a bit on aiming for default dtype dependent behave

- because this is not practical for cases where mixed dtype promotion
rules are not fixed by the standard

steppi committed:
Merge remote-tracking branch 'origin/xp-testing-docs' into xp-testing-docs

steppi committed:
MAINT: remove sphinx-togglebutton from docs requirements txt

steppi commented:
Thanks @rgommers. That should do it hopefully.

rgommers commented:
CI is happy now. Last topic (I think): CI & dev interface. The added `spin check-xp-untested` doesn't look good to me, and I also don't think we want it in CircleCI as a separate job (that makes it take twice longer and is harder to maintain). I can find a place in a good CI job, it's a two-liner to run this as part of an existing job.

For `spin`, here is what it looks like now:
https://github.com/scipy/scipy/pull/24130/files/e0d4b1aa6eeff75b37e47832dea1b0fc062a990b..217812ef73b845e5a3cd3922c88a0ed5c36099b3

It sticks out as quite niche, and not something that the average contributor should care about; I think it'd be better to include as either a `lint` sub-task, or as a new `spin check --xp-markers` sub-task. That would also give a place to add more checks (e.g., the scripts for `Python.h` include order and the symbol hiding checks that are also in `tools/`) without bloating the top-level `spin --help`.

I'd be happy to contribute the last bit in a separate PR if that's easier; then we can just drop the `spin` integration here and get this in. 

WDYT @steppi?

steppi commented:
Thanks @rgommers. That all sounds much better than what I have currently and it would be great if you could submit a follow-up PR to make things nicer. I'll drop the spin interface and take `check-xp-untested` out of CI, but keep the script in `scipy/tools`.

steppi committed:
MAINT: remove spin command and ci job for check-xp-untested

steppi committed:
DOC: temporarily remove docs about what is now called check-xp-untested

steppi committed:
MAINT: Finish removing check_xp_untested CI job

lucascolley committed:
[skip ci]
